features/netsocket/TCPSocket.cpp

```diff
@@ -188,14 +188,8 @@ nsapi_size_or_error_t TCPSocket::recv(void *data, nsapi_size_t size)
 
 void TCPSocket::event()
 {
-    int32_t wcount = _write_sem.wait(0);
-    if (wcount <= 1) {
-        _write_sem.release();
-    }
-    int32_t rcount = _read_sem.wait(0);
-    if (rcount <= 1) {
-        _read_sem.release();
-    }
+    _write_sem.release();
+    _read_sem.release();
```

Comment:
Ah, the lines below do the actual event coalescing. As is this will _technically_ work.

However, it does waste a lot of cycles, since the read/send loop will get woken up redundantly. But we now have the `max_count` parameter on semaphores! So we can actually use a binary semaphore here and the threads will be woken up an optimal number of times.